internal/bufferedread/download_task.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufferedread
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/googlecloudplatform/gcsfuse/v3/internal/block"
+	"github.com/googlecloudplatform/gcsfuse/v3/internal/logger"
+	"github.com/googlecloudplatform/gcsfuse/v3/internal/storage/gcs"
+	"github.com/googlecloudplatform/gcsfuse/v3/internal/workerpool"
+)
+
+type DownloadTask struct {
+	workerpool.Task
+	object *gcs.MinObject
+	bucket gcs.Bucket
+
+	// block is the block to which the data will be downloaded.
+	block block.Block
+
+	// ctx is the context for the download task. It is used to cancel the download.
+	ctx context.Context
+
+	// Used for zonal bucket to bypass the auth & metadata checks.
+	readHandle []byte
+}
+
+func NewDownloadTask(ctx context.Context, object *gcs.MinObject, bucket gcs.Bucket, block block.Block, readHandle []byte) *DownloadTask {
+	return &DownloadTask{
+		ctx:        ctx,
+		object:     object,
+		bucket:     bucket,
+		block:      block,
+		readHandle: readHandle,
+	}
+}
+
+// Execute implements the workerpool.Task interface. It downloads the data from
+// the GCS object to the block.
+// After completion, it notifies the block consumer about the status of the
+// download task. The status can be one of the following:
+// - BlockStatusDownloaded: The download was successful.
+// - BlockStatusDownloadFailed: The download failed due to an error.
+// - BlockStatusDownloadCancelled: The download was cancelled due to context cancellation.
+func (p *DownloadTask) Execute() {
+	startOff := p.block.AbsStartOff()
+	blockId := startOff / p.block.Cap()
+	logger.Tracef("Download: <- block (%s, %v).", p.object.Name, blockId)
+	stime := time.Now()
+	var err error
+	defer func() {
+		if err != nil {
+			logger.Tracef("Download: -> block (%s, %v) failed with error: %v.", p.object.Name, blockId, err)
+		} else {
+			logger.Tracef("Download: -> block (%s, %v): %v completed.", p.object.Name, blockId, time.Since(stime))
```

Comment:
nit: block (%s, %v) completed in : %v

Reply:
Keeping the time for now, as it provides easy way to know the block download time. This will be useful while debug at the start. But, later consider removing this as part of profiler check.